File: python/sudoku.py
# -*- coding: utf-8 -*-
class Sudoku:

    width = 9
    height = 9
    # -1 means empty
    board = [ [-1 for x in range(width)] for y in range(height)]
    empty = width * height

    def __init__(self, board):
        self.height = len(board)
        self.width = len(board[0])
        self.board = board
        
        self.empty = 0
        for y in range(len(board)):
            for x in range(len(board[y])):
                if board[y][x] == -1:
                    self.empty = self.empty +1
        
    def clearPosition(self, x, y):
        self.board[y][x] = -1
        self.empty = self.empty + 1

    def setPosition(self, x, y, value):
        self.board[y][x] = value
        self.empty = self.empty - 1

    def getPosition(self, x, y):
        return self.board[y][x]

    def isValueValid(self, x, y, value):
        
        # check line
        for i in range(self.width):
            if i == x:
                continue
            if value == self.getPosition(i,y):
                return False

        # check column
        for j in range(self.height):
            if j == y:
                continue
            if value == self.getPosition(x,j):
                return False

        # check sub grid (0-3) (3-6) (6-9)
        sectionX = 3*(x//3)
        sectionY = 3*(y//3)
        for i in range(sectionX, sectionX+3):
            for j in range(sectionY, sectionY+3):
                if (x == i) and (y == j):
                    continue
                if value == self.getPosition(i,j):
                    return False
        
        return True

    # test validity, but not completion
    def isValid(self):

        for x in range(self.width):
            for y in range(self.height):
                value = self.getPosition(x, y)
                if value == -1:
                    continue
                if self.isValueValid(x,y, value) == False:
                    return False

        return True

    def computeValidValuesForPosition(self, x, y):

        value = self.getPosition(x, y)
        if value != -1:
            return []
            
        valid = []
        for candidate in range(1, 10):
            if self.isValueValid(x, y, candidate):
                valid.append(candidate)

        return valid        
    
    def solve(self):

        if self.isValid() == False:
            print('Invalid start solution for solve!')
            return

        try:
            self.recursiveSolve()
            print('No solution found.')
        except:
            print('Solution found !')
            print(self.board)
            pass

        

    def recursiveSolve(self):

        if self.isValid() == False:
            return

        if self.empty == 0:
            raise Exception('Found')
        
        # find next most constrained position (candidate)
        candidatePosition = (0,0)
        numberOfValidValues = 9
        valid = []
        for x in range(self.width):
            for y in range(self.height):
                validValues = self.computeValidValuesForPosition(x,y)
                if len(validValues) == 0:
                    continue
                if len(validValues) < numberOfValidValues:
                    candidatePosition = (x,y)
                    numberOfValidValues = len(validValues)
                    valid = validValues
        
        for value in valid:
            self.setPosition(candidatePosition[0], candidatePosition[1], value)
            self.recursiveSolve()
            self.clearPosition(candidatePosition[0], candidatePosition[1])


if __name__ == '__main__':
    
    grille = Sudoku([   [-1, -1, -1, -1, -1, -1, -1, -1, -1 ],
                        [9, 1, 4, 5, 7, -1, -1, 2, 6 ],
                        [5, 2, 8, 6, -1, 1, 4, 9, -1 ],
                        [-1, -1, -1, -1, -1, 3, 2, -1, 4 ],
                        [-1, 6, 3, 1, -1, 9, 5, 7, -1 ],
                        [8, -1, 1, 2, -1, -1, -1, -1, 3 ],
                        [-1, 5, 2, 7, -1, 4, -1, -1, 9 ],
                        [6, 3, -1, 9, -1, -1, 8, 4, -1 ],
                        [4, -1, -1, 3, 2, -1, -1, 5, -1 ]
                     ])

    import datetime
    start = datetime.datetime.now()
    grille.solve()
    end = datetime.datetime.now()
    duration = (end-start).total_seconds() * 1000 * 1000
    print('Solved in '+str(duration)+ ' µs')
    

File: go/sudoku.go
package main

import (
	"fmt"
	"time"
)

var width, height int
var empty int
var values [][]int

func initBoard(val [][]int) {
	width = len(val[0])
	height = len(val)
	values = val
	empty = 0

	for y := 0; y < height; y++ {
		for x := 0; x < width; x++ {
			if values[y][x] == -1 {
				empty = empty + 1
			}
		}
	}
}

func clearPosition(x, y int) {
	values[y][x] = -1
	empty = empty + 1
}

func setPosition(x, y, value int) {
	values[y][x] = value
	empty = empty - 1
}

func getPosition(x, y int) int {
	return values[y][x]
}

func isValueValid(x, y, value int) bool {

	// check line
	for i := 0; i < width; i++ {
		if i == x {
			continue
		}
		if value == getPosition(i, y) {
			return false
		}
	}

	// check column
	for j := 0; j < height; j++ {
		if j == y {
			continue
		}
		if value == getPosition(x, j) {
			return false
		}
	}

	// check sub grid (0-3) (3-6) (6-9)
	sectionX := 3 * (x / 3)
	sectionY := 3 * (y / 3)
	for i := sectionX; i < sectionX+3; i++ {
		for j := sectionY; j < sectionY+3; j++ {
			if (x == i) && (y == j) {
				continue
			}
			if value == getPosition(i, j) {
				return false
			}
		}
	}
	return true
}

// test validity, but not completion
func isValid() bool {

	for x := 0; x < width; x++ {
		for y := 0; y < height; y++ {
			value := getPosition(x, y)
			if value == -1 {
				continue
			}
			if isValueValid(x, y, value) == false {
				return false
			}
		}
	}

	return true
}

func computeValidValuesForPosition(x, y int) []int {

	value := getPosition(x, y)
	valid := []int{}

	if value == -1 {
		for candidate := 1; candidate < 10; candidate++ {
			if isValueValid(x, y, candidate) {
				valid = append(valid, candidate)
			}
		}
	}

	return valid
}

func solve() {

	if isValid() == false {
		fmt.Println("Invalid start solution for solve!")
		return
	}

	result := recursiveSolve()
	if result {
		fmt.Println("Solution found !")
		for line := 0; line < height; line++ {
			fmt.Println(values[line])
		}
	} else {
		fmt.Println("No solution found.")
	}
}

func recursiveSolve() bool {

	if empty == 0 {
		return true
	}

	// find next most constrained position (candidate)
	candidateX, candidateY := 0, 0
	numberOfValidValues := 9
	valid := []int{}
	for x := 0; x < width; x++ {
		for y := 0; y < height; y++ {
			validValues := computeValidValuesForPosition(x, y)
			if len(validValues) == 0 {
				continue
			}
			if len(validValues) < numberOfValidValues {
				candidateX = x
				candidateY = y
				numberOfValidValues = len(validValues)
				valid = validValues
			}
		}
	}

	for _, value := range valid {
		setPosition(candidateX, candidateY, value)
		result := recursiveSolve()
		if result {
			return true
		}
		clearPosition(candidateX, candidateY)
	}

	return false
}

func timeTrack(start time.Time, name string) {
	elapsed := time.Since(start)
	fmt.Printf("%s took %s", name, elapsed)
	recover()
}

func main() {

	//initBoard([][]int{
	//	{-1, -1, -1, -1, -1, -1, -1, -1, -1},
	//	{9, 1, 4, 5, 7, -1, -1, 2, 6},
	//	{5, 2, 8, 6, -1, 1, 4, 9, -1},
	//	{-1, -1, -1, -1, -1, 3, 2, -1, 4},
	//	{-1, 6, 3, 1, -1, 9, 5, 7, -1},
	//	{8, -1, 1, 2, -1, -1, -1, -1, 3},
	//	{-1, 5, 2, 7, -1, 4, -1, -1, 9},
	//	{6, 3, -1, 9, -1, -1, 8, 4, -1},
	//	{4, -1, -1, 3, 2, -1, -1, 5, -1}})

	initBoard([][]int{
		{8, -1, -1, -1, -1, -1, -1, -1, -1},
		{-1, -1, 3, 6, -1, -1, -1, -1, -1},
		{-1, 7, -1, -1, 9, 1, 2, -1, -1},
		{-1, 5, -1, -1, -1, 7, -1, -1, -1},
		{-1, -1, -1, -1, 4, 5, 7, -1, -1},
		{-1, -1, -1, 1, -1, -1, -1, 3, -1},
		{-1, -1, 1, -1, -1, -1, -1, 6, 8},
		{-1, -1, 8, 5, -1, -1, -1, 1, -1},
		{-1, 9, -1, -1, -1, -1, 4, -1, -1}})

	defer timeTrack(time.Now(), "solve")
	solve()
}
